File: wallet/src/main/java/com/worldventures/wallet/service/command/offline_mode/SwitchOfflineModeCommand.java
package com.worldventures.wallet.service.command.offline_mode;

import com.worldventures.janet.injection.InjectableAction;
import com.worldventures.wallet.domain.entity.record.Record;
import com.worldventures.wallet.domain.storage.disk.RecordsStorage;
import com.worldventures.wallet.service.SmartCardInteractor;
import com.worldventures.wallet.service.WalletNetworkService;
import com.worldventures.wallet.service.nxt.DetokenizeMultipleRecordsCommand;
import com.worldventures.wallet.service.nxt.NxtInteractor;
import com.worldventures.wallet.service.nxt.TokenizeMultipleRecordsCommand;
import com.worldventures.wallet.util.NetworkUnavailableException;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import rx.Observable;

@CommandAction
public class SwitchOfflineModeCommand extends Command<Boolean> implements InjectableAction {

   @Inject SmartCardInteractor smartCardInteractor;
   @Inject NxtInteractor nxtInteractor;
   @Inject WalletNetworkService networkService;
   @Inject RecordsStorage recordsStorage;

   @Override
   protected void run(CommandCallback<Boolean> callback) throws Throwable {
      boolean offlineModeEnabled = recordsStorage.readOfflineModeState();

      List<Record> storedRecords = recordsStorage.readRecords();
      if (!storedRecords.isEmpty()) {
         if (!networkService.isAvailable()) {
            throw new NetworkUnavailableException();
         }

         Observable.just(storedRecords)
               .flatMap(records -> offlineModeEnabled ? tokenizeRecords(records) : detokenizeRecords(records))
               .doOnNext(processedRecords -> recordsStorage.saveRecords(processedRecords))
               .flatMap(processedRecords -> smartCardInteractor.offlineModeStatusPipe()
                     .createObservableResult(OfflineModeStatusCommand.save(!offlineModeEnabled)))
               .map(Command::getResult)
               .subscribe(callback::onSuccess, callback::onFail);
      } else {
         smartCardInteractor.offlineModeStatusPipe()
               .createObservableResult(OfflineModeStatusCommand.save(!offlineModeEnabled))
               .map(Command::getResult)
               .subscribe(callback::onSuccess, callback::onFail);
      }
   }

   private Observable<List<Record>> tokenizeRecords(List<? extends Record> records) {
      return nxtInteractor.tokenizeMultipleRecordsPipe()
            .createObservableResult(new TokenizeMultipleRecordsCommand((List<Record>) records, false))
            .map(Command::getResult);
   }

   private Observable<List<Record>> detokenizeRecords(List<? extends Record> records) {
      return nxtInteractor.detokenizeMultipleRecordsPipe()
            .createObservableResult(new DetokenizeMultipleRecordsCommand((List<Record>) records, false))
            .map(Command::getResult);
   }

}

File: wallet/src/main/java/com/worldventures/wallet/service/command/offline_mode/SwitchOfflineModeCommand.kt
package com.worldventures.wallet.service.command.offline_mode

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.domain.entity.record.Record
import com.worldventures.wallet.domain.storage.disk.RecordsStorage
import com.worldventures.wallet.service.SmartCardInteractor
import com.worldventures.wallet.service.WalletNetworkService
import com.worldventures.wallet.service.nxt.DetokenizeMultipleRecordsCommand
import com.worldventures.wallet.service.nxt.NxtInteractor
import com.worldventures.wallet.service.nxt.TokenizeMultipleRecordsCommand
import com.worldventures.wallet.util.NetworkUnavailableException
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import rx.Observable
import javax.inject.Inject

@CommandAction
class SwitchOfflineModeCommand : Command<Boolean>(), InjectableAction {

   @Inject lateinit var smartCardInteractor: SmartCardInteractor
   @Inject lateinit var nxtInteractor: NxtInteractor
   @Inject lateinit var networkService: WalletNetworkService
   @Inject lateinit var recordsStorage: RecordsStorage

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Boolean>) {
      val offlineModeEnabled = recordsStorage.readOfflineModeState()

      val storedRecords = recordsStorage.readRecords()
      if (!storedRecords.isEmpty()) {
         if (!networkService.isAvailable) {
            throw NetworkUnavailableException()
         }

         Observable.just(storedRecords)
               .flatMap { records -> if (offlineModeEnabled) tokenizeRecords(records) else detokenizeRecords(records) }
               .doOnNext { recordsStorage.saveRecords(it) }
               .flatMap {
                  smartCardInteractor.offlineModeStatusPipe()
                        .createObservableResult(OfflineModeStatusCommand.save(!offlineModeEnabled))
               }
               .map { it.result }
               .subscribe({ callback.onSuccess(it) }, { callback.onFail(it) })
      } else {
         smartCardInteractor.offlineModeStatusPipe()
               .createObservableResult(OfflineModeStatusCommand.save(!offlineModeEnabled))
               .map{ it.result }
               .subscribe( { callback.onSuccess(it) }, { callback.onFail(it) })
      }
   }

   private fun tokenizeRecords(records: List<Record>): Observable<List<Record>> {
      return nxtInteractor.tokenizeMultipleRecordsPipe()
            .createObservableResult(TokenizeMultipleRecordsCommand(records, false))
            .map{ it.result }
   }

   private fun detokenizeRecords(records: List<Record>): Observable<List<Record>> {
      return nxtInteractor.detokenizeMultipleRecordsPipe()
            .createObservableResult(DetokenizeMultipleRecordsCommand(records, false))
            .map{ it.result }
   }

}
